src/groupeddataframe/grouping.jl

```diff
@@ -177,6 +213,318 @@ end
+may be `NamedTuple` (default) or `Tuple`.
+
+### Returns
+
+Iterator over `Tuple`s/`NamedTuple`s, a single `Tuple`/`NamedTuple`, or a
+grouping column value.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(groupvalues(gd))
+6-element Array{NamedTuple{(:a, :b),Tuple{Symbol,Int64}},1}:
+ (a = :foo, b = 2)
+ (a = :bar, b = 1)
+ (a = :baz, b = 2)
+ (a = :foo, b = 1)
+ (a = :bar, b = 2)
+ (a = :baz, b = 1)
+
+julia> groupvalues(gd, 2)
+(a = :bar, b = 1)
+
+julia> groupvalues(gd, 2, 1)
+:bar
+
+julia> groupvalues(gd, 2, :a)
+:bar
+
+julia> groupvalues(Tuple, gd, 2)
+(:bar, 1)
+```
+"""
+groupvalues(T::Type, gd::GroupedDataFrame) = (groupvalues(T, gd, i) for i in 1:length(gd))
+groupvalues(gd::GroupedDataFrame) = groupvalues(NamedTuple, gd)
+
+groupvalues(::Type{Tuple}, gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
+groupvalues(::Type{NamedTuple}, gd::GroupedDataFrame, i::Integer) = NamedTuple{Tuple(groupvars(gd))}(groupvalues(Tuple, gd, i))
+groupvalues(gd::GroupedDataFrame, i::Integer) = groupvalues(NamedTuple, gd, i)
+
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Integer) = gd.parent[gd.idx[gd.starts[i]], gd.cols[col]]
+groupvalues(gd::GroupedDataFrame, i::Integer, col::Symbol) = groupvalues(gd, i, _groupvar_idx(gd, col, true))
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+
+"""
+    GroupKey{T<:GroupedDataFrame}
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{T<:GroupedDataFrame}
+    parent::T
+    idx::Int
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey: ")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.keys(key::GroupKey) = Tuple(groupvars(parent(key)))
```

Comment:
Same here, why not just return the vector?
```suggestion
Base.keys(key::GroupKey) = groupvars(parent(key))
```

Reply:
This was for consistency with `NamedTuple`.

Reply:
Yeah, I realize we also do that for `DataFrameRow`, but I wonder whether there's any advantage to that.

@bkamins Do you have an opinion?

Reply:
The only advantage is consistency with `NamedTuple` other than that there is no benefit. This is not a performance critical function so I think having it consistent with `NamedTuple` is OK.